# TP introduction à **PySpark** et **DBT**

auteur: César Gattano

**Rappel du brief:**
Comprendre comment utiliser **PySpark** pour transformer des données massives et **DBT** pour modéliser ces données dans un entrepôt.

Vous travaillez pour une société de ventes à distance. Votre mission est de :
- Nettoyer et transformer des données avec **PySpark**.
- Charger ces données dans une base **PostgreSQL**.
- Construire un modèle en étoile simple avec **DBT**.
- Répondre à certaines questions métier via **DBT**.

**Dataset:** [ventes.csv](https://drive.google.com/file/d/1kzNj66aWbtnOYZIwTj2nux-knLp2bETs/view?usp=drive_link) \
Cet ensemble de données contient des informations sur les ventes réalisées par une entreprise, chaque ligne correspond à une vente.

*Voir aussi le [Notion](https://www.notion.so/Spark-Pyspark-27d7bf1b1fa180419888f91f351aec50?source=copy_link) pour en savoir plus sur **Spark** et **PySpark**, et le [Notion](https://www.notion.so/DBT-27e7bf1b1fa180c5bd7dd8ee7d87db63?source=copy_link) pour en savoir plus sur **DBT***.


## Nettoyer et transformer des données avec **PySpark**

### Initialisation de la Session Spark

In [1]:
from pyspark.sql import SparkSession

spark = SparkSession.builder.getOrCreate()

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
25/10/02 11:01:28 WARN Utils: Your hostname, gattano-ThinkPad-P53, resolves to a loopback address: 127.0.1.1; using 192.168.1.130 instead (on interface enx00e04c680450)
25/10/02 11:01:28 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
25/10/02 11:01:29 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
25/10/02 11:01:30 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.


### Chargement des données depuis le CSV

In [2]:
ventes_schema = \
    'id_transaction INT, ' + \
    'client_nom STRING, ' + \
    'client_age INT, ' + \
    'client_ville STRING, ' + \
    'produit_nom STRING, ' + \
    'produit_categorie STRING, ' + \
    'produit_marque STRING, ' + \
    'prix_catalogue INT, ' + \
    'magasin_nom STRING, ' + \
    'magasin_type STRING, ' + \
    'magasin_region STRING,' + \
    'date DATE, ' + \
    'quantite INT, ' + \
    'montant_total INT'

ventes_00 = spark.read.csv('data/ventes.csv', schema=ventes_schema, header=True)

Configuration pour éviter de devoir appeler `.show()` dans ce notebook.

In [3]:
spark.conf.set('spark.sql.repl.eagerEval.enabled', True)
spark.conf.set('spark.sql.repl.eagerEval.maxNumRows', 10)
ventes_00

id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,date,quantite,montant_total
1,Alice,25,Paris,Ordinateur,Informatique,Dell,800,Boutique Lyon,Physique,Auvergne-Rhône-Alpes,2023-03-12,2,NULL
2,Bob,34,Lyon,Smartphone,Téléphonie,Apple,1200,Boutique Lyon,Physique,Auvergne-Rhône-Alpes,2023-01-27,5,NULL
3,Alice,25,Paris,Montre connectée,Accessoires,Garmin,300,E-Shop,En ligne,National,2023-01-09,1,NULL
4,Alice,25,Paris,Smartphone,Téléphonie,Apple,1200,Boutique Paris,Physique,Île-de-France,2023-05-10,5,NULL
5,Alice,25,Paris,Montre connectée,Accessoires,Garmin,300,Boutique Paris,Physique,Île-de-France,2023-06-16,5,NULL
6,David,40,Bordeaux,Smartphone,Téléphonie,Apple,1200,E-Shop,En ligne,National,2023-05-31,3,NULL
7,Alice,25,Paris,Smartphone,Téléphonie,Apple,1200,Boutique Lyon,Physique,Auvergne-Rhône-Alpes,2023-04-19,3,NULL
8,Charlie,29,Marseille,Smartphone,Téléphonie,Apple,1200,Boutique Paris,Physique,Île-de-France,2023-03-28,1,NULL
9,Alice,25,Paris,Tablette,Informatique,Samsung,600,Boutique Paris,Physique,Île-de-France,2023-04-02,3,NULL
10,Emma,31,Toulouse,Casque audio,Accessoires,Sony,150,Boutique Paris,Physique,Île-de-France,2023-04-28,5,NULL


Statistiques générales

In [4]:
ventes_00.describe()

25/10/02 11:02:32 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,495,495,495,495,495,495,495,495,495,495,495,487,0
mean,248.0,45.0,130.16363636363636,NULL,NULL,3.0,NULL,596.2626262626262,3158.5,NULL,NULL,2.975359342915811,NULL
stddev,143.03845636750978,46.66904755831214,2191.127734403582,NULL,NULL,0.0,NULL,407.8694169120773,1678.5361678160725,NULL,NULL,1.3831022506046147,NULL
min,1,12,-29,Bordeaux,Casque audio,3,Apple,-1200,3547,En ligne,Auvergne-Rhône-Alpes,1,NULL
max,495,Emma,48781,Toulouse,Tablette,Téléphonie,Sony,1200,E-Shop,Physique,Île-de-France,5,NULL


### Mise en place du pipeline de transformation

#### **1. Mettre tous les champs strings en lowercase**

In [5]:
from pyspark.sql.functions import lower

ventes_01 = ventes_00. \
    withColumn('client_nom', lower(ventes_00.client_nom)). \
    withColumn('client_ville', lower(ventes_00.client_ville)). \
    withColumn('produit_nom', lower(ventes_00.produit_nom)). \
    withColumn('produit_categorie', lower(ventes_00.produit_categorie)). \
    withColumn('produit_marque', lower(ventes_00.produit_marque)). \
    withColumn('magasin_nom', lower(ventes_00.magasin_nom)). \
    withColumn('magasin_type', lower(ventes_00.magasin_type)). \
    withColumn('magasin_region', lower(ventes_00.magasin_region))
ventes_01

id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,date,quantite,montant_total
1,alice,25,paris,ordinateur,informatique,dell,800,boutique lyon,physique,auvergne-rhône-alpes,2023-03-12,2,NULL
2,bob,34,lyon,smartphone,téléphonie,apple,1200,boutique lyon,physique,auvergne-rhône-alpes,2023-01-27,5,NULL
3,alice,25,paris,montre connectée,accessoires,garmin,300,e-shop,en ligne,national,2023-01-09,1,NULL
4,alice,25,paris,smartphone,téléphonie,apple,1200,boutique paris,physique,île-de-france,2023-05-10,5,NULL
5,alice,25,paris,montre connectée,accessoires,garmin,300,boutique paris,physique,île-de-france,2023-06-16,5,NULL
6,david,40,bordeaux,smartphone,téléphonie,apple,1200,e-shop,en ligne,national,2023-05-31,3,NULL
7,alice,25,paris,smartphone,téléphonie,apple,1200,boutique lyon,physique,auvergne-rhône-alpes,2023-04-19,3,NULL
8,charlie,29,marseille,smartphone,téléphonie,apple,1200,boutique paris,physique,île-de-france,2023-03-28,1,NULL
9,alice,25,paris,tablette,informatique,samsung,600,boutique paris,physique,île-de-france,2023-04-02,3,NULL
10,emma,31,toulouse,casque audio,accessoires,sony,150,boutique paris,physique,île-de-france,2023-04-28,5,NULL


#### **2. Éliminer les lignes où une donnée est manquante**

Par exemple:

In [6]:
ventes_01.filter("quantite IS NULL")

id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,date,quantite,montant_total
12,emma,31,toulouse,casque audio,accessoires,sony,150,boutique lyon,physique,auvergne-rhône-alpes,2023-02-19,NULL,NULL
58,emma,31,toulouse,tablette,informatique,samsung,600,boutique paris,physique,île-de-france,2023-02-18,NULL,NULL
64,emma,31,toulouse,montre connectée,accessoires,garmin,300,boutique paris,physique,île-de-france,2023-06-08,NULL,NULL
90,alice,25,paris,smartphone,téléphonie,apple,1200,e-shop,en ligne,national,2023-03-26,NULL,NULL
110,bob,34,lyon,tablette,informatique,samsung,600,boutique lyon,physique,auvergne-rhône-alpes,2023-01-19,NULL,NULL
117,emma,31,toulouse,tablette,informatique,samsung,600,boutique lyon,physique,auvergne-rhône-alpes,2023-04-25,NULL,NULL
363,emma,31,toulouse,smartphone,téléphonie,apple,1200,e-shop,en ligne,national,2023-01-24,NULL,NULL
446,bob,34,lyon,casque audio,accessoires,sony,150,boutique lyon,physique,auvergne-rhône-alpes,2023-04-04,NULL,NULL


In [7]:
import pandas as pd

def pandas_drop_partial_entries(iterator):
    for pandas_df in iterator:
        yield pandas_df[
            (pd.notnull(pandas_df.id_transaction)) &
            (pd.notnull(pandas_df.client_nom)) &
            (pd.notnull(pandas_df.client_age)) &
            (pd.notnull(pandas_df.client_ville)) &
            (pd.notnull(pandas_df.produit_nom)) &
            (pd.notnull(pandas_df.produit_categorie)) &
            (pd.notnull(pandas_df.produit_marque)) &
            (pd.notnull(pandas_df.prix_catalogue)) &
            (pd.notnull(pandas_df.magasin_nom)) &
            (pd.notnull(pandas_df.magasin_type)) &
            (pd.notnull(pandas_df.magasin_region)) &
            (pd.notnull(pandas_df.date)) &
            (pd.notnull(pandas_df.quantite))
        ]
ventes_02 = ventes_01.mapInPandas(pandas_drop_partial_entries, schema=ventes_01.schema)

In [8]:
ventes_02.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,487,487,487,487,487,487,487,487,487,487,487,487,0
mean,249.48665297741272,45.0,131.79260780287476,NULL,NULL,3.0,NULL,596.2012320328543,3158.5,NULL,NULL,2.975359342915811,NULL
stddev,142.48610980778824,46.66904755831214,2209.050834314869,NULL,NULL,0.0,NULL,408.1612695660718,1678.5361678160725,NULL,NULL,1.3831022506046147,NULL
min,1,12,-29,bordeaux,casque audio,3,apple,-1200,3547,en ligne,auvergne-rhône-alpes,1,NULL
max,495,emma,48781,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


Méthode alternative à privilégier:

In [9]:
ventes_02 = ventes_01.dropna(subset=(
    'id_transaction',
    'client_nom',
    'client_age',
    'client_ville',
    'produit_nom',
    'produit_categorie',
    'produit_marque',
    'prix_catalogue',
    'magasin_nom',
    'magasin_type',
    'magasin_region',
    'date',
    'quantite'
))

In [10]:
ventes_02.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,487,487,487,487,487,487,487,487,487,487,487,487,0
mean,249.48665297741272,45.0,131.79260780287476,NULL,NULL,3.0,NULL,596.2012320328543,3158.5,NULL,NULL,2.975359342915811,NULL
stddev,142.48610980778824,46.66904755831214,2209.050834314869,NULL,NULL,0.0,NULL,408.1612695660718,1678.5361678160725,NULL,NULL,1.3831022506046147,NULL
min,1,12,-29,bordeaux,casque audio,3,apple,-1200,3547,en ligne,auvergne-rhône-alpes,1,NULL
max,495,emma,48781,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** 8 entrées ont été supprimées

#### **3. Éliminer les lignes où la date d'achat est extravagante**

In [11]:
ventes_02.select("date").sort("date", ascending=True)

date
1632-03-07
1702-06-28
1742-01-16
1745-04-27
1821-06-25
1852-04-14
2023-01-01
2023-01-01
2023-01-02
2023-01-03


In [12]:
ventes_02.select("date").sort("date", ascending=False)

date
2023-06-29
2023-06-29
2023-06-29
2023-06-29
2023-06-29
2023-06-28
2023-06-28
2023-06-27
2023-06-27
2023-06-27


Les dates aberrantes sont concentrées avant 2023-01-01.

In [14]:
ventes_03 = ventes_02.where(ventes_02.date >= '2023-01-01')
ventes_03.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,481,481,481,481,481,481,481,481,481,481,481,481,0
mean,249.32016632016632,45.0,133.04989604989606,NULL,NULL,3.0,NULL,596.1538461538462,3158.5,NULL,NULL,2.972972972972973,NULL
stddev,142.85573065149416,46.66904755831214,2222.785537901758,NULL,NULL,0.0,NULL,407.6108862817388,1678.5361678160725,NULL,NULL,1.3834201644299364,NULL
min,1,12,-29,bordeaux,casque audio,3,apple,-1200,3547,en ligne,auvergne-rhône-alpes,1,NULL
max,495,emma,48781,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** 6 entrées ont été supprimées

#### **4. Si l'âge est négatif, le mettre en positif + éliminez les entrées avec un âge aberrant**

In [15]:
ventes_03.select("client_age").sort("client_age", ascending=True)

client_age
-29
-25
25
25
25
25
25
25
25
25


In [16]:
ventes_03.select("client_age").sort("client_age", ascending=False)

client_age
48781
40
40
40
40
40
40
40
40
40


In [17]:
from pyspark.sql.functions import abs

ventes_04 = ventes_03.where(ventes_03.client_age < 100).withColumn('client_age', abs(ventes_03.client_age))
ventes_04.describe()


summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,480,480,480,480,480,480,480,480,480,480,480,480,0
mean,249.39166666666668,45.0,31.925,NULL,NULL,3.0,NULL,596.7708333333334,3158.5,NULL,NULL,2.972916666666667,NULL
stddev,142.99615574103876,46.66904755831214,4.9795615673393145,NULL,NULL,0.0,NULL,407.81124310577735,1678.5361678160725,NULL,NULL,1.3848629309461271,NULL
min,1,12,25,bordeaux,casque audio,3,apple,-1200,3547,en ligne,auvergne-rhône-alpes,1,NULL
max,495,emma,40,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** 1 entrée a été supprimée et 2 entrée ont été transformées

#### **5. Gérer les noms de clients aberrants**

In [18]:
ventes_04.select("client_nom").sort("client_nom", ascending=True).distinct()

client_nom
david
alice
78
charlie
emma
12
bob


In [19]:
ventes_04.filter(ventes_04.client_nom.isin(['12', '78']))

id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,date,quantite,montant_total
146,78,31,toulouse,ordinateur,informatique,dell,800,e-shop,en ligne,national,2023-05-17,4,NULL
210,12,40,bordeaux,tablette,informatique,samsung,600,789,physique,île-de-france,2023-04-27,5,NULL


L'entrée avec le nom de client "12" a aussi un nom de magasin qui semble aberrant ("789"). On va supprimer cette entrée.
L'entrée avec le nom de client "78" semble cohérente pour le reste. On va renommer ce client en 'inconnu'.

In [20]:
from pyspark.sql.functions import udf

@udf(returnType='string')
def name_or_unknown(name: str):
    if name == '78':
        return 'inconnu'
    else:
        return name

ventes_05 = ventes_04.where(ventes_04.client_nom != '12').withColumn('client_nom', name_or_unknown(ventes_04["client_nom"]))
ventes_05.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,479,479,479,479,479,479,479,479,479,479,479,479,0
mean,249.47390396659708,NULL,31.908141962421713,NULL,NULL,3.0,NULL,596.7640918580375,3948.3333333333335,NULL,NULL,2.9686847599164925,NULL
stddev,143.1342920924348,NULL,4.971037087006122,NULL,NULL,0.0,NULL,408.2375742558096,695.1297241043095,NULL,NULL,1.3832003635063412,NULL
min,1,alice,25,bordeaux,casque audio,3,apple,-1200,3547,en ligne,auvergne-rhône-alpes,1,NULL
max,495,inconnu,40,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** 1 entrée a été supprimée et 1 entrée a été transformée

#### **6. Gérer les noms de ville aberrantes**

In [21]:
ventes_05.select("client_ville").distinct()

client_ville
bordeaux
paris
lyon
marseille
toulouse


**Résultat:** Pas besoin de transformation

#### **7. Gérer les noms de produits aberrants**

In [22]:
ventes_05.select("produit_nom").distinct()

produit_nom
smartphone
montre connectée
ordinateur
tablette
casque audio


**Résultat:** Pas besoin de transformation

#### **8. Gérer les catégories de produits aberrantes**

In [23]:
ventes_05.select("produit_nom","produit_categorie").distinct()

produit_nom,produit_categorie
smartphone,3
smartphone,téléphonie
ordinateur,3
ordinateur,informatique
montre connectée,accessoires
tablette,informatique
casque audio,accessoires


In [24]:
ventes_05.createOrReplaceTempView("ventes_05")
ventes_08 = spark.sql("""SELECT
            id_transaction,
            client_nom,
            client_age,
            client_ville,
            produit_nom,
            CASE
                WHEN produit_categorie == '3' AND produit_nom == 'smartphone' THEN "téléphonie"
                WHEN produit_categorie == '3' AND produit_nom == 'ordinateur' THEN "informatique"
                WHEN produit_categorie == '3' THEN '3'
                ELSE produit_categorie
            END AS produit_categorie,
            produit_marque,
            prix_catalogue,
            magasin_nom,
            magasin_type,
            magasin_region,
            date,
            quantite,
            montant_total
          FROM ventes_05""")
ventes_08.select("produit_nom","produit_categorie").distinct()


produit_nom,produit_categorie
smartphone,téléphonie
ordinateur,informatique
montre connectée,accessoires
tablette,informatique
casque audio,accessoires


In [25]:
ventes_08.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,479,479,479,479,479,479,479,479,479,479,479,479,0
mean,249.47390396659708,NULL,31.908141962421713,NULL,NULL,NULL,NULL,596.7640918580375,3948.3333333333335,NULL,NULL,2.9686847599164925,NULL
stddev,143.1342920924348,NULL,4.971037087006122,NULL,NULL,NULL,NULL,408.2375742558096,695.1297241043095,NULL,NULL,1.3832003635063412,NULL
min,1,alice,25,bordeaux,casque audio,accessoires,apple,-1200,3547,en ligne,auvergne-rhône-alpes,1,NULL
max,495,inconnu,40,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** 5 entrées ont été transformées

#### **9. Gérer les marques de produits aberrantes**

In [26]:
ventes_08.select("produit_marque").distinct()

produit_marque
apple
dell
sony
samsung
garmin


**Résultat:** Pas besoin de transformation

#### **10. Gérer les prix_catalogue aberrant**

In [27]:
ventes_08.sort("prix_catalogue", ascending=True)

id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,date,quantite,montant_total
448,emma,31,toulouse,smartphone,téléphonie,apple,-1200,e-shop,en ligne,national,2023-05-22,3,NULL
20,alice,25,paris,smartphone,téléphonie,apple,-850,boutique lyon,physique,auvergne-rhône-alpes,2023-03-22,2,NULL
136,david,40,bordeaux,ordinateur,informatique,dell,-800,boutique lyon,physique,auvergne-rhône-alpes,2023-05-19,1,NULL
282,emma,31,toulouse,tablette,informatique,samsung,-600,boutique paris,physique,île-de-france,2023-06-29,1,NULL
318,alice,25,paris,tablette,informatique,samsung,-600,boutique paris,physique,île-de-france,2023-03-04,4,NULL
377,emma,31,toulouse,tablette,informatique,samsung,-600,boutique paris,physique,île-de-france,2023-06-20,1,NULL
188,charlie,29,marseille,montre connectée,accessoires,garmin,-300,boutique lyon,physique,auvergne-rhône-alpes,2023-05-26,3,NULL
374,emma,31,toulouse,montre connectée,accessoires,garmin,-300,boutique paris,physique,île-de-france,2023-06-11,5,NULL
83,charlie,29,marseille,casque audio,accessoires,sony,-150,boutique paris,physique,île-de-france,2023-04-02,2,NULL
78,david,40,bordeaux,casque audio,accessoires,sony,150,boutique paris,physique,île-de-france,2023-01-01,3,NULL


In [28]:
ventes_08.select("produit_nom", "produit_marque", "prix_catalogue").distinct().sort("produit_nom")

produit_nom,produit_marque,prix_catalogue
casque audio,sony,150
casque audio,sony,-150
montre connectée,garmin,-300
montre connectée,garmin,300
ordinateur,dell,-800
ordinateur,dell,800
smartphone,apple,-1200
smartphone,apple,-850
smartphone,apple,1200
tablette,samsung,-600


Mise à part des montants qui semblent être passés en négatif, il n'y pas de valeur vraiment aberrante. Peut-être que le smartphone Apple à 850 pourrait être étudié plus précisément puisqu'il semble être une combinaison d'entrée unique dans la base de donnée...

In [29]:
from pyspark.sql.functions import abs

ventes_10 = ventes_08.withColumn('prix_catalogue', abs(ventes_08.prix_catalogue))
ventes_10.describe()


summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,479,479,479,479,479,479,479,479,479,479,479,479,0
mean,249.47390396659708,NULL,31.908141962421713,NULL,NULL,NULL,NULL,619.3110647181628,3948.3333333333335,NULL,NULL,2.9686847599164925,NULL
stddev,143.1342920924348,NULL,4.971037087006122,NULL,NULL,NULL,NULL,373.07069296374067,695.1297241043095,NULL,NULL,1.3832003635063412,NULL
min,1,alice,25,bordeaux,casque audio,accessoires,apple,150,3547,en ligne,auvergne-rhône-alpes,1,NULL
max,495,inconnu,40,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** 9 entrées ont été transformées

#### **11. Gérer les informations de magasins aberrantes**

In [30]:
ventes_10.select("magasin_nom", "magasin_type", "magasin_region").distinct()

magasin_nom,magasin_type,magasin_region
boutique lyon,physique,auvergne-rhône-alpes
e-shop,en ligne,national
3547,physique,auvergne-rhône-alpes
4751,physique,île-de-france
boutique paris,physique,île-de-france


In [31]:
ventes_10.createOrReplaceTempView("ventes_10")
ventes_11 = spark.sql("""SELECT
            id_transaction,
            client_nom,
            client_age,
            client_ville,
            produit_nom,
            produit_categorie,
            produit_marque,
            prix_catalogue,
            CASE
                WHEN magasin_nom == "3547" AND magasin_type == "physique" AND magasin_region == "auvergne-rhône-alpes" THEN "boutique lyon"
                WHEN magasin_nom == "4751" AND magasin_type == "physique" AND magasin_region == "île-de-france" THEN "boutique paris"
                ELSE magasin_nom
            END AS magasin_nom,
            magasin_type,
            magasin_region,
            date,
            quantite,
            montant_total
          FROM ventes_10""")
ventes_11.select("magasin_nom","magasin_type", "magasin_region").distinct()

magasin_nom,magasin_type,magasin_region
boutique lyon,physique,auvergne-rhône-alpes
e-shop,en ligne,national
boutique paris,physique,île-de-france


In [32]:
ventes_11.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,479,479,479,479,479,479,479,479,479,479,479,479,0
mean,249.47390396659708,NULL,31.908141962421713,NULL,NULL,NULL,NULL,619.3110647181628,NULL,NULL,NULL,2.9686847599164925,NULL
stddev,143.1342920924348,NULL,4.971037087006122,NULL,NULL,NULL,NULL,373.07069296374067,NULL,NULL,NULL,1.3832003635063412,NULL
min,1,alice,25,bordeaux,casque audio,accessoires,apple,150,boutique lyon,en ligne,auvergne-rhône-alpes,1,NULL
max,495,inconnu,40,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** 2 entrées ont été transformées

#### **12. Gérer les quantités aberrantes**

In [33]:
ventes_11.select("quantite").distinct().sort("quantite", ascending=True)

quantite
1
2
3
4
5


**Résultat:** Pas besoin de transformation.

#### **13. Recherche de doublons**

In [34]:
ventes_12 = ventes_11.dropDuplicates(
    ['client_nom',
    'client_age',
    'client_ville',
    'produit_nom',
    'produit_categorie',
    'produit_marque',
    'prix_catalogue',
    'magasin_nom',
    'magasin_type',
    'magasin_region',
    'date',
    'quantite'])
ventes_12.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,478,478,478,478,478,478,478,478,478,478,478,478,0
mean,249.44769874476987,NULL,31.90376569037657,NULL,NULL,NULL,NULL,620.2928870292887,NULL,NULL,NULL,2.9686192468619246,NULL
stddev,143.28309923047954,NULL,4.975321323074022,NULL,NULL,NULL,NULL,372.84154208571664,NULL,NULL,NULL,1.3846487560295841,NULL
min,1,alice,25,bordeaux,casque audio,accessoires,apple,150,boutique lyon,en ligne,auvergne-rhône-alpes,1,NULL
max,495,inconnu,40,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,NULL


**Résultat:** une entrée a été supprimée.

#### **13. Calculer le montant total**

In [35]:
ventes_clean = ventes_12.withColumn('montant_total', ventes_12.prix_catalogue * ventes_12.quantite)
ventes_clean.describe()

summary,id_transaction,client_nom,client_age,client_ville,produit_nom,produit_categorie,produit_marque,prix_catalogue,magasin_nom,magasin_type,magasin_region,quantite,montant_total
count,478,478,478,478,478,478,478,478,478,478,478,478,478
mean,249.44769874476987,NULL,31.90376569037657,NULL,NULL,NULL,NULL,620.2928870292887,NULL,NULL,NULL,2.9686192468619246,1829.2887029288702
stddev,143.28309923047954,NULL,4.975321323074022,NULL,NULL,NULL,NULL,372.84154208571664,NULL,NULL,NULL,1.3846487560295841,1478.4971392227035
min,1,alice,25,bordeaux,casque audio,accessoires,apple,150,boutique lyon,en ligne,auvergne-rhône-alpes,1,150
max,495,inconnu,40,toulouse,tablette,téléphonie,sony,1200,e-shop,physique,île-de-france,5,6000


## Charger ces données dans une base **PostgreSQL**

### Chargement des variables d'environnement pour Postgres

In [36]:
import os
from dotenv import load_dotenv

load_dotenv(".env.postgres")
for env_key in os.environ:
    print(f"{env_key:40}: {os.environ[env_key]}")

BAMF_DESKTOP_FILE_HINT                  : /var/lib/snapd/desktop/applications/code_code.desktop
CHROME_DESKTOP                          : code.desktop
CLUTTER_DISABLE_MIPMAPPED_TEXT          : 1
DBUS_SESSION_BUS_ADDRESS                : unix:path=/run/user/1000/bus
DEBUGINFOD_URLS                         : https://debuginfod.ubuntu.com 
DESKTOP_SESSION                         : ubuntu
DISPLAY                                 : :1
ELECTRON_NO_ATTACH_CONSOLE              : 1
FONTCONFIG_FILE                         : /etc/fonts/fonts.conf
FONTCONFIG_PATH                         : /etc/fonts
GDK_BACKEND                             : x11
GDK_PIXBUF_MODULEDIR                    : /snap/code/208/usr/lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders
GDK_PIXBUF_MODULE_FILE                  : /home/gattano/snap/code/common/.cache/gdk-pixbuf-loaders.cache
GDMSESSION                              : ubuntu
GIO_LAUNCHED_DESKTOP_FILE               : /var/lib/snapd/desktop/applications/code_code.deskto

### Création des types de données

In [ ]:
import psycopg

URI_str = f"postgresql://{os.environ["POSTGRES_USER"]}:{os.environ["POSTGRES_PASSWORD"]}@localhost:{os.environ["LOCAL_DB_PORT"]}/{os.environ["POSTGRES_DB"]}"
print(URI_str)

conn = psycopg.connect(URI_str)
with conn.cursor() as cur:
    # Creates enumerations
    cur.execute("""
        CREATE TYPE type_de_magasin AS ENUM (
                'physique',
                'en ligne'
            )
        """)
    cur.execute("""
        CREATE TYPE nom_de_magasin AS ENUM (
                'boutique lyon',
                'boutique paris',
                'e-shop'
            )
        """)
    cur.execute("""
        CREATE TYPE region AS ENUM (
                'auvergne-rhône-alpes',
                'île-de-france',
                'national'
            )
        """)
    cur.execute("""
        CREATE TYPE categorie_de_produit AS ENUM (
                'téléphonie',
                'informatique',
                'accessoires'
            )
        """)

conn.commit()
conn.close()

postgresql://admin:admin@localhost:5432/intro_pyspark_dbt-db


### Création de la table de données

In [41]:
import psycopg

URI_str = f"postgresql://{os.environ["POSTGRES_USER"]}:{os.environ["POSTGRES_PASSWORD"]}@localhost:{os.environ["LOCAL_DB_PORT"]}/{os.environ["POSTGRES_DB"]}"
print(URI_str)

conn = psycopg.connect(URI_str)
with conn.cursor() as cur:
    # Creates the new table
    cur.execute("""
        CREATE TABLE IF NOT EXISTS ventes (
            id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
            id_transaction integer NOT NULL UNIQUE,
            client_nom varchar(30) NOT NULL,
            client_age smallint NOT NULL CHECK (client_age > 0),
            client_ville varchar(30) NOT NULL,
            produit_nom varchar(30) NOT NULL,
            produit_categorie categorie_de_produit NOT NULL,
            produit_marque varchar(30) NOT NULL,
            prix_catalogue smallint NOT NULL CHECK (prix_catalogue > 0),
            magasin_nom nom_de_magasin NOT NULL,
            magasin_type type_de_magasin NOT NULL,
            magasin_region region NOT NULL,
            date date NOT NULL CHECK (date >= '2023-01-01'),
            quantite smallint NOT NULL CHECK (quantite > 0),
            montant_total integer NOT NULL CHECK (montant_total > 0)
            )
        """)

conn.commit()
conn.close()

postgresql://admin:admin@localhost:5432/intro_pyspark_dbt-db


### Insertion des données

In [42]:
import psycopg

URI_str = f"postgresql://{os.environ["POSTGRES_USER"]}:{os.environ["POSTGRES_PASSWORD"]}@localhost:{os.environ["LOCAL_DB_PORT"]}/{os.environ["POSTGRES_DB"]}"
print(URI_str)

conn = psycopg.connect(URI_str)
with conn.cursor() as cur:

    for row in ventes_clean.collect():
        row.asDict()
        # TODO: Inefficient way to store mutliple data... Search for COPY or PREPARE in psycopg doc or JDBC from spark tools
        cur.execute("""
            INSERT INTO ventes VALUES (
                DEFAULT,
                %(id_transaction)s,
                %(client_nom)s,
                %(client_age)s,
                %(client_ville)s,
                %(produit_nom)s,
                %(produit_categorie)s,
                %(produit_marque)s,
                %(prix_catalogue)s,
                %(magasin_nom)s,
                %(magasin_type)s,
                %(magasin_region)s,
                %(date)s,
                %(quantite)s,
                %(montant_total)s
                )
        """, row.asDict())
conn.commit()
conn.close()


postgresql://admin:admin@localhost:5432/intro_pyspark_dbt-db


TODO: Alternative pour l'insertion des données: explorez l'utilisation de JDBC (regardez des tutos)

## Construire un modèle en étoile simple avec **DBT**

### Configuration de **DBT**

La commande `dbt init` a été utilisée

#### Fichier `dbt_project.yml`

```YAML
name: "intro_pyspark_dbt"
profile: "intro_pyspark_dbt"
models:
  intro_pyspark_dbt:
    +materialized: view
    stagged:
      +materialized: view
      +schema: stagged
    marts:
      +materialized: view
      +schema: marts
    analytics:
      +materialized: view
      +schema: analytics
vars:
  "dbt_date:time_zone": "Europe/Paris"
```

#### Fichier `packages.yml`

```YAML
packages:
  - package: dbt-labs/dbt_utils
    version: 1.3.1
```

#### Fichier `profiles.yml`

```YAML
intro_pyspark_dbt:
  target: dev
  outputs:
    dev:
      type: postgres
      host: localhost
      user: admin
      password: ...
      port: 5432
      dbname: intro_pyspark_dbt-db
      schema: dbt_cgattano
      threads: 4
      keepalives_idle: 0
      connect_timeout: 10
      retries: 1

### Stagged modèle des ventes

```SQL
with stagged_ventes as (
    SELECT
        id,
        id_transaction,
        client_nom,
        client_age,
        client_ville,
        produit_nom,
        produit_categorie,
        produit_marque,
        prix_catalogue,
        magasin_nom,
        magasin_type,
        magasin_region,
        date,
        quantite,
        montant_total
    FROM ventes
)

SELECT * FROM stagged_ventes
```

Question: Est-ce que cette vue est réellement nécessaire puisque la mise en base d'une table a déjà été faite au préalable en python ?

### Modèles des dimensions

#### Dimension client

Fichier `dim_client.sql`

```SQL
WITH ventes AS (
    SELECT * FROM {{ ref('stg_ventes') }}
),
final AS (
    SELECT
        row_number() over () AS id,
        client_nom AS nom,
        client_age AS age,
        client_ville AS ville
    FROM ventes
    GROUP BY client_nom, client_age, client_ville
    ORDER BY client_nom, client_age, client_ville
)

SELECT * FROM final
```

Fichier `dim_client.yml`

```YAML
version: 2

models:
  - name: dim_client
    description: Une entrée pour chaque client
    columns:
      - name: id
        description: La clé primaire d'identification du client
        data_tests:
          - not_null
          - unique
      - name: nom
        description: Le nom du client (obligatoire)
        data_tests:
          - not_null
      - name: age
        description: L'âge du client (obligatoire, > 0)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 0
                inclusive: false
      - name: ville
        description: La ville du client (obligatoire)
        data_tests:
          - not_null
```

#### Dimention magasin

Fichier `dim_magasin.sql`

```SQL
WITH ventes AS (
    SELECT * FROM {{ ref('stg_ventes') }}
),
final AS (
    SELECT
        row_number() over () AS id,
        magasin_nom AS nom,
        magasin_type AS type,
        magasin_region AS region
    FROM ventes
    GROUP BY nom, type, region
    ORDER BY type, region, nom
)

SELECT * FROM final
```

Fichier `dim_magasin.yml`

```YAML
version: 2

models:
  - name: dim_magasin
    description: Une entrée pour chaque magasin
    columns:
      - name: id
        description: La clé primaire d'identification du magasin
        data_tests:
          - not_null
          - unique
      - name: nom
        description: Le nom du magasin (obligatoire)
        data_tests:
          - not_null
      - name: type
        description: Le type de magasin (obligatoire, enum)
        data_tests:
          - not_null
          - accepted_values:
              arguments:
                values: ["physique", "en ligne"]
      - name: region
        description: La région du magasin (obligatoire)
        data_tests:
          - not_null
```

#### Dimension produit

Fichier `dim_produit.sql`

```SQL
WITH ventes AS (
    SELECT * FROM {{ ref('stg_ventes') }}
),
final AS (
    SELECT
        row_number() over () AS id,
        produit_nom AS nom,
        produit_categorie AS categorie,
        produit_marque AS marque
    FROM ventes
    GROUP BY nom, categorie, marque
    ORDER BY categorie, marque, nom
)

SELECT * FROM final
```

Fichier `dim_produit.yml`

```YAML
version: 2

models:
  - name: dim_produit
    description: Une entrée pour chaque produit
    columns:
      - name: id
        description: La clé primaire d'identification du produit
        data_tests:
          - not_null
          - unique
      - name: nom
        description: Le nom du produit (obligatoire)
        data_tests:
          - not_null
      - name: categorie
        description: La catégorie du produit (obligatoire, enum)
        data_tests:
          - not_null
          - accepted_values:
              arguments:
                values: ["informatique", "accessoires", "téléphonie"]
      - name: marque
        description: La marque du client (obligatoire)
        data_tests:
          - not_null
```

#### Dimension temps

Fichier `dim_temps.sql`

```SQL
WITH ventes AS (
    SELECT date FROM {{ ref('stg_ventes') }}
),
final AS (
    SELECT
        date,
        EXTRACT(YEAR FROM date) AS year,
        EXTRACT(MONTH FROM date) AS month,
        EXTRACT(ISOYEAR FROM date) AS isoyear,
        EXTRACT(WEEK FROM date) AS week,
        EXTRACT(DAY FROM date) AS day,
        EXTRACT(DOW FROM date) AS dow,
        EXTRACT(DOY FROM date) AS doy
    FROM ventes
    GROUP BY date
    ORDER BY date
)

SELECT * FROM final
```

Fichier `dim_temps.yml`

```YAML
version: 2

models:
  - name: dim_temps
    description: Une entrée pour chaque date avec au moins un achat
    columns:
      - name: date
        description: La date du jour d'un achat
        data_tests:
          - not_null
          - unique
      - name: year
        description: L'année (obligatoire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 2023
                max_value: 2025
                inclusive: true
      - name: month
        description: Le mois (obligatoire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 1
                max_value: 12
                inclusive: true
      - name: day
        description: Le jour (obligatoire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 1
                max_value: 31
                inclusive: true
      - name: dow
        description: Le jour de la semaine (obligatoire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 0
                max_value: 6
                inclusive: true
      - name: doy
        description: Le jour de l'année (obligatoire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 1
                max_value: 366
                inclusive: true
      - name: week
        description: La semaine (obligatoire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 1
                max_value: 53
                inclusive: true
      - name: isoyear
        description: L'année ISO associée au numéro de semaine (obligatoire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 2022
                max_value: 2025
                inclusive: true
```

### Modèle de la table des faits

Fichier `fct_ventes.sql`

```SQL
WITH ventes AS (
    SELECT * FROM {{ ref('stg_ventes') }}
),
dim_temps AS (
    SELECT date FROM {{ ref('dim_temps') }}
),
dim_client AS (
    SELECT * FROM {{ ref('dim_client') }}
),
dim_magasin AS (
    SELECT * FROM {{ ref('dim_magasin') }}
),
dim_produit AS (
    SELECT * FROM {{ ref('dim_produit') }}
),
final AS (
    SELECT
        v.id AS id,
        id_transaction,
        c.id AS id_client,
        p.id AS id_produit,
        m.id AS id_magasin,
        v.date AS date,
        quantite,
        prix_catalogue AS prix_unitaire,
        montant_total
    FROM ventes AS v
    INNER JOIN dim_temps AS t
    ON v.date = t.date
    INNER JOIN dim_client AS c
    ON v.client_nom = c.nom AND v.client_age = c.age AND v.client_ville = c.ville
    INNER JOIN dim_produit AS p
    ON v.produit_nom = p.nom AND v.produit_categorie = p.categorie AND v.produit_marque = p.marque
    INNER JOIN dim_magasin AS m
    ON v.magasin_nom = m.nom AND v.magasin_type = m.type AND v.magasin_region = m.region
)

SELECT * FROM final
```

Fichier `fct_ventes.yml`

```YAML
version: 2

models:
  - name: fct_ventes
    description: Une entrée pour chaque achat d'une référence produit en plusieurs quantités
    columns:
      - name: id
        description: La clé primaire d'identification d'un achat
        data_tests:
          - not_null
          - unique
      - name: id_transaction
        description: La clé d'identification d'une transaction
        data_tests:
          - not_null
      - name: id_client
        description: Le clé d'identification du client (obligatoire)
        data_tests:
          - not_null
      - name: id_produit
        description: Le clé d'identification du produit (obligatoire)
        data_tests:
          - not_null
      - name: id_magasin
        description: Le clé d'identification du magasin (obligatoire)
        data_tests:
          - not_null
      - name: date
        description: La date d'achat (obligatoire)
        data_tests:
          - not_null
      - name: quantite
        description: La quantité du produit acheté (obligatoire, > 0)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 0
                inclusive: false
      - name: prix_unitaire
        description: Le prix catalogue du produit au moment de l'achat (obligatoire, > 0)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 0
                inclusive: false
      - name: montant_total
        description: Le montant total de la commande (quantite * prix_unitaire)
        data_tests:
          - not_null
          - dbt_utils.accepted_range:
              arguments:
                min_value: 0
                inclusive: false
```

## Répondre aux questions métier via **DBT**

### Analyse temporelle


#### 1.1 Quel est le chiffre d’affaires mensuel de l'entreprise ? 

Requête:

```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_temps AS (
    SELECT * FROM {{ ref('dim_temps') }}
)
SELECT
    case
        when month = 1 then 'janvier'
        when month = 2 then 'février'        
        when month = 3 then 'mars'        
        when month = 4 then 'avril'        
        when month = 5 then 'mai'        
        when month = 6 then 'juin'        
        when month = 7 then 'juillet'        
        when month = 8 then 'août'        
        when month = 9 then 'septembre'        
        when month = 10 then 'octobre'        
        when month = 11 then 'novembre'          
        when month = 12 then 'decembre'       
    end as month,
    SUM(montant_total) AS ca,
    COUNT(id_transaction) AS nb_transactions,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) / COUNT(id_transaction) AS panier_moyen
FROM fact_ventes AS v
INNER JOIN dim_temps AS t
ON v.date = t.date
GROUP BY month
ORDER BY ca DESC
```

Réponse: 
| Mois | CA |
| :--- | :- |
| janvier | 151550 |
| février | 109250 |
| mars    | 165200 |
| avril   | 135450 |
| mai     | 184550 |
| juin    | 128400 |


#### Quels sont les mois les plus rentables de l’année ? (Faire Top 3)


Requête: même requête que ci-dessus

Réponse: 
| Mois | CA |
| :--- | :- |
| mai     | 184550 |
| mars    | 165200 |
| janvier | 151550 |


#### Quelle est l’évolution du panier moyen mensuel (montant_total / transaction) au fil du temps ?

Requête: même requête que ci-dessus

Réponse: 
| Mois | panier_moyen |
| :--- | :- |
| janvier | 1782 |
| février | 1680 |
| mars    | 1966 |
| avril   | 1593 |
| mai     | 2073 |
| juin    | 1834 |

### Analyse produit


#### Quels sont les produits les plus vendus en quantité ?

Requête:

```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_produit AS (
    SELECT * FROM {{ ref('dim_produit') }}
)
SELECT
    ANY_VALUE(p.nom) AS nom,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_produit AS p
ON v.id_produit = p.id
GROUP BY p.id
ORDER BY ca DESC
```

Réponse: 
| produit | nb_vendus |
| :--- | :- |
| montre connectée   | 333 |
| smartphone | 297 |
| tablette    | 295 |
| ordinateur | 258 |
| casque audio     | 236 |



#### Quels produits génèrent le plus de chiffre d’affaires ?

Requête: même requête que ci-dessus

Réponse: 
| produit | ca |
| :--- | :- |
| smartphone | 355700 |
| ordinateur | 206400 |
| tablette    | 177000 |
| montre connectée   | 99900 |
| casque audio     | 35400 |



#### Quelles catégories de produits sont les plus populaires ?

Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_produit AS (
    SELECT * FROM {{ ref('dim_produit') }}
)
SELECT
    p.categorie,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_produit AS p
ON v.id_produit = p.id
GROUP BY p.categorie
ORDER BY ca DESC
```

Réponse: 
| categorie | nb_vendus |
| :--- | :- |
| informatique | 553 |
| téléphonie | 297 |
| accessoires    | 569 |

### Analyse client


#### Quelle est la répartition des ventes par tranche d’âge (10 ans) ?

Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_client AS (
    SELECT * FROM {{ ref('dim_client') }}
)
SELECT
    CONCAT(
        CAST((c.age / 10)*10 AS CHAR(2)),
        '-', 
        CAST((c.age / 10)*10+9 AS CHAR(2))
    ) AS tranche_age,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_client AS c
ON v.id_client = c.id
GROUP BY tranche_age
ORDER BY ca DESC
```

Réponse: 
| tranche_age | nb_vendus | ca |
| :--- | :- | :- |
| 20-29 | 566 | 359700 |
| 30-39 | 575 | 346050 |
| 40-49 | 278 | 168650 |


#### Quelle ville génère le plus de ventes ?

Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_client AS (
    SELECT * FROM {{ ref('dim_client') }}
)
SELECT
    ville,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_client AS c
ON v.id_client = c.id
GROUP BY ville
ORDER BY ca DESC
```

Réponse: 
| ville | nb_vendus | ca |
| :--- | :- | :- |
| toulouse | 332 | 193700 |
| marseille | 305 | 186000 |
| paris | 261 | 173700 |
| bordeaux | 278 | 168650 |
| lyon | 243 | 152350 |


### Bonus : 


#### Analyse magasin / canal de vente

Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_magasin AS (
    SELECT * FROM {{ ref('dim_magasin') }}
)
SELECT
    ANY_VALUE(m.nom) AS nom,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_magasin AS m
ON v.id_magasin = m.id
GROUP BY m.id
ORDER BY ca DESC
```

Réponse: 
| nom | nb_vendus | ca |
| :--- | :- | :- |
| boutique paris | 510 | 303600 |
| e-shop | 441 | 287850 |
| boutique lyon | 468 | 282950 |

Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_magasin AS (
    SELECT * FROM {{ ref('dim_magasin') }}
)
SELECT
    m.type AS type,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_magasin AS m
ON v.id_magasin = m.id
GROUP BY m.type
ORDER BY ca DESC
```

Réponse: 
| type | nb_vendus | ca |
| :--- | :- | :- |
| physique | 978 | 586550 |
| en ligne | 441 | 287850 |


#### Quelle part de vente (nombres de ventes) vient des boutiques physiques vs e-shop ?


Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_magasin AS (
    SELECT * FROM {{ ref('dim_magasin') }}
)
SELECT
    m.type AS type,
    CAST(
        COUNT(id_transaction)*100.0 / (SELECT COUNT(*) FROM fact_ventes) 
        AS SMALLINT
    ) AS perc_ventes
FROM fact_ventes AS v
INNER JOIN dim_magasin AS m
ON v.id_magasin = m.id
GROUP BY m.type
```

Réponse: 
| type | perc_ventes |
| :--- | :- |
| physique | 68 |
| en ligne | 32 |


#### Quel magasin physique génère le plus de transactions ?

Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_magasin AS (
    SELECT * FROM {{ ref('dim_magasin') }}
)
SELECT
    ANY_VALUE(m.nom) AS nom,
    COUNT(id_transaction) AS nb_transac
FROM fact_ventes AS v
INNER JOIN dim_magasin AS m
ON v.id_magasin = m.id
WHERE m.type = 'physique'
GROUP BY m.id
ORDER BY nb_transac DESC
```

Réponse: 
| nom | nb_transac |
| :--- | :- |
| boutique lyon | 162 |
| boutique paris | 162 |


### Analyse croisée


#### Quels produits sont achetés par les jeunes clients (<30 ans) ? (faire un classement, quantités vendues et non nombre de ventes)


Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_client AS (
    SELECT * FROM {{ ref('dim_client') }}
),
dim_produit AS (
    SELECT * FROM {{ ref('dim_produit') }}
)
SELECT
    ANY_VALUE(p.nom) AS produit,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_client AS c
ON v.id_client = c.id
INNER JOIN dim_produit AS p
ON v.id_produit = p.id
WHERE c.age < 30
GROUP BY p.id
ORDER BY nb_vendus DESC
```

Réponse: 
| produit | nb_vendus | ca |
| :--- | :- | :- |
| smartphone | 132 | 157700 |
| tablette | 123 | 73800 |
| montre connectée | 119 | 35700 |
| ordinateur | 98 | 78400 |
| casque audio | 94 | 14100 |


#### Quelles catégories sont les plus vendues sur le e-shop vs en boutique ?


Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_magasin AS (
    SELECT * FROM {{ ref('dim_magasin') }}
),
dim_produit AS (
    SELECT * FROM {{ ref('dim_produit') }}
)
SELECT
    m.type AS type_magasin,
    p.categorie AS categorie_produit,
    SUM(quantite) AS nb_vendus,
    SUM(montant_total) AS ca
FROM fact_ventes AS v
INNER JOIN dim_magasin AS m
ON v.id_magasin = m.id
INNER JOIN dim_produit AS p
ON v.id_produit = p.id
GROUP BY m.type, p.categorie
ORDER BY m.type, nb_vendus DESC
```

Réponse: 
| type_magasin | categorie_produit | nb_vendus | ca |
| :--- | :- | :- | :- |
| phyisque | accessoires | 415 | 98250 |
| physique | informatique | 361 | 246600 |
| physique | téléphonie | 202 | 241700 |
| en ligne | accessoires | 192 | 136800 |
| en ligne | informatique | 154 | 37050 |
| en ligne | téléphonie | 95 | 114000 |


#### Existe-t-il une saisonnalité (ex: pics sur certains mois/produits) ?

Requête:
```SQL
WITH fact_ventes AS (
    SELECT * FROM {{ ref('fct_ventes') }}
),
dim_temps AS (
    SELECT * FROM {{ ref('dim_temps') }}
),
dim_produit AS (
    SELECT * FROM {{ ref('dim_produit') }}
)
SELECT
    month,
    case
        when month = 1 then 'janvier'
        when month = 2 then 'février'        
        when month = 3 then 'mars'        
        when month = 4 then 'avril'        
        when month = 5 then 'mai'        
        when month = 6 then 'juin'        
        when month = 7 then 'juillet'        
        when month = 8 then 'août'        
        when month = 9 then 'septembre'        
        when month = 10 then 'octobre'        
        when month = 11 then 'novembre'          
        when month = 11 then 'decembre'       
    end as nom_month,
    ANY_VALUE(p.nom) AS produit,
    COUNT(id_transaction) AS nb_transac,
    SUM(quantite) AS nb_vendus
FROM fact_ventes AS v
INNER JOIN dim_temps AS t
ON v.date = t.date
INNER JOIN dim_produit AS p
ON v.id_produit = p.id
GROUP BY p.id, month
ORDER BY produit, month
```

Réponse: trop long à retranscrire...